### PR TITLE
HW Barrier

### PR DESCRIPTION
**Pull request by tommydcjung**

### CODE REVIEW

File: software/spmd/hw_barrier_16x8_interrupt_test/interrupt.S [hunk 1]
```diff
@@ -0,0 +1,107 @@
+#include "bsg_manycore_arch.h"
```
Comment (taylor-bsg) [inline software/spmd/hw_barrier_16x8_interrupt_test/interrupt.S:1]: description of test rationale; what does it intended to test, and what are the expected behaviors.

File: software/spmd/hw_barrier_16x8_interrupt_test/main.c [hunk 2]
```diff
@@ -0,0 +1,109 @@
+
```
Comment (taylor-bsg) [inline software/spmd/hw_barrier_16x8_interrupt_test/main.c:1]: "

File: software/spmd/hw_barrier_NxN_test/main.c [hunk 3]
```diff
@@ -0,0 +1,95 @@
+#include "bsg_manycore.h"
```
Comment (taylor-bsg) [inline software/spmd/hw_barrier_NxN_test/main.c:1]: "

File: software/spmd/hw_barrier_context_switch_test/interrupt.S [hunk 4]
```diff
@@ -0,0 +1,223 @@
+#include "bsg_manycore_arch.h"
```
Comment (taylor-bsg) [inline software/spmd/hw_barrier_context_switch_test/interrupt.S:1]: "

File: software/spmd/hw_barrier_context_switch_test/main.c [hunk 5]
```diff
@@ -0,0 +1,143 @@
+#include "bsg_manycore.h"
```
Comment (taylor-bsg) [inline software/spmd/hw_barrier_context_switch_test/main.c:1]: "

File: v/vanilla_bean/cl_decode.v [hunk 6]
```diff
@@ -123,7 +123,19 @@ assign decode_o.is_jal_op = instruction_i.op == `RV32_JAL_OP;
...
+  decode_o.is_fence_op = 1'b0;
+  decode_o.is_barsend_op = 1'b0;
+  decode_o.is_barrecv_op = 1'b0;
+  if (instruction_i ==? `RV32_FENCE_OP) begin
+    // fence FM
+    unique casez (instruction_i[31:28])
+      `RV32_FENCE_FM: decode_o.is_fence_op = 1'b1;
+      `RV32_BARSEND_FM: decode_o.is_barsend_op = 1'b1;
+      `RV32_BARRECV_FM: decode_o.is_barrecv_op = 1'b1;
+    endcase
```
Comment (taylor-bsg) [inline v/vanilla_bean/cl_decode.v:136]: missing default

**Review by taylor-bsg — COMMENTED** [review]
In addition to annotated suggestions; will probably need to rebase to incorporate the `BSG_INVALID_PARAM stuff; and then update in this module to use that macro as well

File: software/bsg_manycore_lib/bsg_barrier_amoadd.S [hunk 7]
```diff
@@ -64,6 +64,7 @@ bsg_barrier_amoadd_sleep:
   lr.w t0, 0(a1)
   beq t3, t0, bsg_barrier_amoadd_end
   lr.w.aq t0, 0(a1)
+  beq t3, t0, bsg_barrier_amoadd_sleep
```
Comment (taylor-bsg) [inline software/bsg_manycore_lib/bsg_barrier_amoadd.S:67]: // we need to check this, in order to support the reservation 
// being cleared by a context switch

File: software/spmd/hw_barrier_16x8_interrupt_test/main.c (hunk 2 above)
Comment (taylor-bsg) [reply software/spmd/hw_barrier_16x8_interrupt_test/main.c:6]: This describes what the test does, but should also try to describe what functionality and corner cases you are intending to cover with this test

Comment (taylor-bsg) [reply software/spmd/hw_barrier_16x8_interrupt_test/main.c:6]: (same for some of the other comments)

File: v/vanilla_bean/cl_decode.v (hunk 6 above)
Comment (taylor-bsg) [reply v/vanilla_bean/cl_decode.v:142]: You can just write default:  with no code, if you want the values preserved from the header initializations.

**Comment by drichmond** [comment]
Where are we on merging this? Anything I can do to help?

**Review by drichmond — APPROVED** [review]
